Make sign tx testable

sevanspowell opened the pull request:
This PR is primarily about improving the testability of signTransaction, and providing a number of property tests.

- [x] In the signTransaction code path, pushed IO further to the edge of the program, enabling us to unit test more of the logic of signTransaction.
- [x] Provided a number of property tests to prove key points of signTransaction (it's witness adding behaviour)
    - [x] Also, specifically added a test to assert that signTransaction maintains the script integrity hash of a transaction, for ADP-1047.
- [x]  Expanded the transaction generator slightly to possibly generate ScriptWitnesses for TxIns.

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-1047 / ADP-1139

sevanspowell committed:
Move IO out of sign transaction and into server handler

- For signTransaction:
  - Move IO operations out of the wallet layer and into the server handler. We
    want to move IO operations to the edge of our program, in this case, the
    "edge" is our server handler. Doing so makes the wallet layer easier to
    test.

sevanspowell committed:
Add property tests for signTransaction

- Add the following property tests for signTransaction:
  - adds the reward account witness when necessary
  - adds extra key witnesses when necessary
  - adds tx in witnesses when necessary
  - adds collateral witnesses when necessary
  - never removes witnesses

sevanspowell committed:
Check script integrity hash

sevanspowell committed:
Expand Tx generator to create script witnesses for TxIns

- This update allows the spending witness for a TxIn to be a script witness or a
  key witness. Previously only key witnesses were supported.

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1878,8 +1881,31 @@ signTransaction ctx (ApiT wid) body = do
-        liftHandler $ W.signTransaction wrk wid pwd sealedTx
+    sealedTx' <- withWorkerCtx ctx wid liftE liftE $ \wrk -> liftHandler $ do
+        let
+            db = wrk ^. W.dbLayer @IO @s @k
```
We have those parts in almost all places in `where`. Shouldn't we have here also (just for consistency)?

sevanspowell replied:
It won't work in this case, because `wrk` is not an argument to the outer function, `wrk` won't be in scope in the `where` block.

paweljakubas commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -281,43 +337,18 @@ import Test.Utils.Paths
 import qualified Cardano.Api as Cardano
+import qualified Cardano.Api.Gen as Cardano
 import qualified Cardano.Api.Shelley as Cardano
+import qualified Cardano.Crypto.Hash.Blake2b as Blake2b
```
what about using here `as Crypto` ?

paweljakubas replied:
or just import without qualification like in other places in codebase?

paweljakubas commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1968,43 +1969,39 @@ selectAssets ctx pp params transform = do
+  -- necessary
+signTransaction tl era keyLookup (rootKey, rootPwd) utxo =
+    let
+        rewardAcnt :: (XPrv, Passphrase "encryption")
```
what about exposing this function? It is used also in tests. Maybe it would be good to have code reuse

paweljakubas reviewed (approved):
very nice properties added. It would be great to continue this path of testing when adding test coverage of decodeTx (in some forthcoming PRs). 
1. We see that signing can add some bits and this is tested.What about properties showing that signing will not change inps, outs, validity, ...., some things' ordering already present in the body?  
2. I see common patterns in properties. For example,  we have the same structure:
```
case Cardano.THINGSupportedInEra era of
    Nothing -> 
    Just _ -> 
```  
I wonder if there is a way to extract that common structure and reuse it in all property tests?

sevanspowell committed:
Extract common "feature supported in era" code

sevanspowell committed:
Import as Crypto for consistency

sevanspowell committed:
Ensure signTransaction doesn't change TxBody at all

- Add tests to ensure signTransaction doesn't modify TxBody.
- Make "checkSubsetOf" test more concise and reusable.

rvl commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -345,6 +376,434 @@ spec = do
+    -- ^ Extra keys to form basis of address -> key lookup function
+    -> Property
+prop_signTransaction_neverChangesTxBody (AnyCardanoEra era) rootK utxo extraKeys =
+    forAll (genTxInEra era) $ \tx -> do
+        let
```
May as well combine the let and where bindings at the deeper level...
```suggestion
    forAll (genTxInEra era) $ \tx -> bodyContentBefore == bodyContentAfter
        where
            tl = testTxLayer
            
```
(same goes for the other `prop_` functions)

rvl commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -345,6 +376,434 @@ spec = do
+        [ x `elem` xs
+        & counterexample ("actual set: " <> show xs)
+        & counterexample ("expected elem: " <> show x)
+        ]
```
I don't think the foldMap is needed?
```suggestion
    counterexample ("actual set: " <> show xs) $ conjoin
        [ x `elem` xs & counterexample ("expected elem: " <> show x)
        | x <- elems ]
```

rvl commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -345,6 +376,434 @@ spec = do
     updateSealedTxSpec
     balanceTransactionSpec
+    describe "Sign transaction" $ do
+        forAllEras (\era -> it ("signTransaction adds reward account witness when necessary (" <> show era <> ")") $
+            property (prop_signTransaction_addsRewardAccountKey era))
```
I was thinking that it would be nice to have the property description/right beside the property definition, but still exposing the `Property` as a standalone function. Something like this...

```suggestion
        describe desc_signTransaction_addsRewardAccountKey $
            forAllEras $ \era -> it (show era) $
                property (prop_signTransaction_addsRewardAccountKey era))
            
        -- ... --
        
desc_signTransaction_addsRewardAccountKey = "signTransaction adds reward account witness when necessary" :: String
```

sevanspowell replied:
Agree, could be cool. But I wouldn't want these to be the only tests that do it.

rvl commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -345,6 +376,434 @@ spec = do
+            rootK = first liftRawKey rootXPrv
+
+            rawRewardK :: (XPrv, Passphrase "encryption")
+            rawRewardK = (getRawKey $ deriveRewardAccount (snd rootK) (fst rootK), snd rootK)
```
If we find that tests are taking too long to run due to key generation, we could shift these things up into a precooked record of key inputs.

rvl commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -345,6 +376,434 @@ spec = do
+                                [mkShelleyWitness txBody rawRewardK]
+
+            checkCoverage $
+                expectedWits `checkSubsetOf` (getSealedTxWitnesses sealedTx')
```
Cool. I was wondering whether the reward account xpub could be used to verify new witnesses, but it seems easier and more effective to just generate the witness directly, as you have done.

rvl reviewed (approved):
Excellent work - these are really good tests to have! :star:

sevanspowell committed:
Combine where and let bindings

sevanspowell committed:
Remove unnecessary foldMap
